File: frontend/src/api/waypoint.ts
// HTTP client wrappers for /api/waypoint/*

export interface Waypoint {
  id: number;
  api_id: string | null;
  lat: number;
  lon: number;
  name: string;
  children: number[];
  visited: boolean;
}

export interface WaypointTree {
  id: number;
  api_id: string | null;
  lat: number;
  lon: number;
  name: string;
  children: WaypointTree[];
  visited: boolean;
}

export async function getWaypoint(id: number): Promise<Waypoint> {
  const res = await fetch(`/api/waypoint/${id}`)
  if (!res.ok) throw new Error(`Failed to fetch waypoint ${id}: ${res.status}`)
  return res.json() as Promise<Waypoint>
}

export async function getWaypointTree(userId: number): Promise<WaypointTree> {
  const res = await fetch(`/api/waypoint/tree/${userId}`)
  if (!res.ok) throw new Error(`Failed to fetch waypoint tree for user ${userId}: ${res.status}`)
  return res.json() as Promise<WaypointTree>
}

export async function setVisited(id: number, visited: boolean = true): Promise<Waypoint> {
  const res = await fetch(`/api/waypoint/${id}/visited`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ visited }),
  })
  if (!res.ok) throw new Error(`Failed to set visited on waypoint ${id}: ${res.status}`)
  return res.json() as Promise<Waypoint>
}

export async function createWaypoint(params: {
  api_id?: string;
  lat: number;
  lon: number;
  name: string;
}): Promise<Waypoint> {
  const res = await fetch('/api/waypoint', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) throw new Error(`Failed to create waypoint: ${res.status}`)
  return res.json() as Promise<Waypoint>
}

export async function addChildren(parentId: number, childIds: number[]): Promise<Waypoint> {
  const res = await fetch(`/api/waypoint/${parentId}/children`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ child_ids: childIds }),
  })
  if (!res.ok) throw new Error(`Failed to add children to waypoint ${parentId}: ${res.status}`)
  return res.json() as Promise<Waypoint>
}

export async function discoverNearby(
  lat: number,
  lon: number,
  rad?: number,
  num?: number,
): Promise<Waypoint[]> {
  const res = await fetch('/api/waypoint/osm', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ lat, lon, rad: rad ?? 500, num: num ?? 3 }),
  })
  if (!res.ok) throw new Error(`Failed to discover nearby waypoints: ${res.status}`)
  return res.json() as Promise<Waypoint[]>
}

export async function exploreWaypoint(
  parentId: number,
  lat: number,
  lon: number,
  rad?: number,
  num?: number,
): Promise<Waypoint> {
  const waypoints = await discoverNearby(lat, lon, rad, num)
  const newIds = waypoints.map(w => w.id)
  return addChildren(parentId, newIds)
}

File: backend/routes/waypoint.py
"""Waypoint API routes."""

from flask import Blueprint, g, jsonify, request, Response
from loguru import logger

from backend.db.waypoint_queries import (
    add_children_to_waypoint,
    create_waypoint,
    get_waypoint as get_waypoint_query,
    get_waypoint_tree_for_user,
    set_waypoint_visited,
)
from backend.services.osm import query_nearby

waypoint_bp = Blueprint("waypoint", __name__, url_prefix="/api/waypoint")


# GET /api/waypoint/<id>
@waypoint_bp.route("/<int:waypoint_id>", methods=["GET"])
def get_waypoint(waypoint_id: int) -> tuple[Response, int]:
    """Return one waypoint by ID."""
    waypoint = get_waypoint_query(g.db, waypoint_id)
    if not waypoint:
        return jsonify({"error": "Waypoint not found"}), 404
    return jsonify(waypoint.to_dict()), 200


# GET /api/waypoint/tree/<user_id>
@waypoint_bp.route("/tree/<int:user_id>", methods=["GET"])
def get_tree_by_user(user_id: int) -> tuple[Response, int]:
    """Return the full waypoint tree for a user."""
    tree = get_waypoint_tree_for_user(g.db, user_id)
    if tree is None:
        return jsonify({"error": "User or root waypoint not found"}), 404
    return jsonify(tree), 200


# PATCH /api/waypoint/<id>/visited
@waypoint_bp.route("/<int:waypoint_id>/visited", methods=["PATCH"])
def set_visited(waypoint_id: int) -> tuple[Response, int]:
    """Update visited status for one waypoint."""
    payload = request.get_json(silent=True) or {}
    visited = bool(payload.get("visited", True))
    waypoint = set_waypoint_visited(g.db, waypoint_id, visited)
    if not waypoint:
        return jsonify({"error": "Waypoint not found"}), 404
    logger.info(
        f"Waypoint {waypoint_id} marked as {'visited' if visited else 'unvisited'}"
    )
    return jsonify(waypoint.to_dict()), 200


# POST /api/waypoint
@waypoint_bp.route("", methods=["POST"])
def create_single_waypoint() -> tuple[Response, int]:
    """Create a single waypoint."""
    payload = request.get_json(silent=True) or {}
    lat = payload.get("lat")
    lon = payload.get("lon")
    name = payload.get("name")
    api_id = payload.get("api_id", None)

    if lat is None or lon is None or name is None or api_id is None:
        return jsonify({"error": "lat, lon, name, and api_id are required"}), 400

    waypoint = create_waypoint(g.db, api_id, lat, lon, name)
    return jsonify(waypoint.to_dict()), 201


# PATCH /api/waypoint/<id>/children
@waypoint_bp.route("/<int:waypoint_id>/children", methods=["PATCH"])
def add_children(waypoint_id: int) -> tuple[Response, int]:
    """Add children to a waypoint."""
    payload = request.get_json(silent=True) or {}
    child_ids = payload.get("child_ids")

    if child_ids is None or not isinstance(child_ids, list):
        return jsonify({"error": "child_ids must be a list"}), 400

    add_children_to_waypoint(g.db, waypoint_id, child_ids)
    waypoint = get_waypoint_query(g.db, waypoint_id)
    if not waypoint:
        return jsonify({"error": "Waypoint not found"}), 404
    logger.info(f"Added {len(child_ids)} children to waypoint {waypoint_id}")
    return jsonify(waypoint.to_dict()), 200


# POST /api/waypoint/osm
@waypoint_bp.route("/osm", methods=["POST"])
def create_from_osm() -> tuple[Response, int]:
    """Query OSM for nearby POIs and create waypoints from the results."""
    payload = request.get_json(silent=True) or {}
    lat = payload.get("lat")
    lon = payload.get("lon")

    if lat is None or lon is None:
        return jsonify({"error": "lat and lon are required"}), 400

    num = payload.get("num", 10)

    results = query_nearby(lat, lon, limit=num)

    created = []
    for r in results:
        waypoint = create_waypoint(
            g.db, api_id=str(r["id"]), lat=r["lat"], lon=r["lon"], name=r["name"]
        )
        created.append(waypoint.to_dict())

    logger.info(
        f"Created {len(created)} waypoints from OSM query at ({lat:.4f}, {lon:.4f})"
    )
    return jsonify(created), 201
